src/sync/browser.js

```diff
@@ -0,0 +1,105 @@
+
+  const contexts = {};
+  let pushManager;
+
+  function startPolling() {
+    log.info('PUSH down or disconnected. Starting periodic fetch of data.');
+    forOwn(contexts, function (context) {
+      const producer = context.get(context.constants.PRODUCER, true);
+      if (producer && !producer.isRunning())
+        producer.start(true); // `fetchers` are scheduled but not called immediately
+    });
+  }
+
+  function stopPollingAndSyncAll() {
+    log.info('PUSH (re)connected. Syncing and stopping periodic fetch of data.');
+    // if polling, stop
+    forOwn(contexts, function (context) {
+      const producer = context.get(context.constants.PRODUCER, true);
+      if (producer && producer.isRunning())
+        producer.stop();
+    });
+    syncAll();
+  }
+
+  function syncAll() {
+    // fetch splits and segments
+    const mainProducer = mainContext.get(mainContext.constants.PRODUCER, true);
+    mainProducer && mainProducer.synchronizeSplits();
+    forOwn(contexts, function (context) {
+      const producer = context.get(context.constants.PRODUCER, true);
+      producer && producer.synchronizeMySegments();
+    });
+  }
+
+  function shared(context, isSharedClient = true) {
+    const producer = isSharedClient ? PartialProducerFactory(context) : FullProducerFactory(context);
+    context.put(context.constants.PRODUCER, producer);
+    const settings = context.get(context.constants.SETTINGS);
+    const userKey = matching(settings.core.key);
+    if(contexts[userKey]) log.warn('A client with the same user key has already been created. Only the new instance will be properly synchronized.');
+    contexts[userKey] = context;
+
+    return {
+      start() {
+        // start syncing
+        if (pushManager) {
+          if (!isSharedClient) {
+            syncAll(); // initial syncAll (only when main client is created)
+            pushManager.on(PUSH_CONNECT, stopPollingAndSyncAll);
+            pushManager.on(PUSH_DISCONNECT, startPolling);
+          }
+          pushManager.startNewClient(userKey, context);
+        } else {
+          producer.start();
+        }
+      },
+
+      stop() {
+        const context = contexts[userKey];
+        if (context) { // check in case `client.destroy()` has been invoked more than once for the same client
+          delete contexts[userKey];
+
+          if (pushManager) {
+            pushManager.removeClient(userKey);
+            // stop push if stoping main client
+            if (!isSharedClient)
+              pushManager.stop();
+            // We don't reconnect pushmanager when removing a shared client,
+            // since it is more costly than continue listening the channel
+          }
+
+          if (producer && producer.isRunning())
+            producer.stop();
+
+        }
+      }
+    };
+  }
+
+  // called before creating PushManager, to create the producer and put in context.
+  const result = shared(mainContext, false);
```

Comment:
"result" is not representative at all. why not just syncManager or mainSyncManager ?